jax/_src/util.py

```diff
@@ -450,22 +450,24 @@ def wrapper(fun: T) -> T:
 def assert_unreachable(x):
   raise AssertionError(f"Unhandled case: {type(x).__name__}")
 
-def tuple_insert(t, idx, val):
```

Comment:
I think once the other suggested changes are implemented, we could revert all changes to `jax/_src/util.py`.